File: src/main/java/gloridifice/watersource/common/recipe/serializer/WaterLevelRecipeSerializer.java
package gloridifice.watersource.common.recipe.serializer;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import gloridifice.watersource.common.recipe.WaterLevelAndEffectRecipe;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.GsonHelper;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.ForgeRegistryEntry;

import java.util.ArrayList;
import java.util.List;

public class WaterLevelRecipeSerializer<T extends WaterLevelAndEffectRecipe> extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<T> {


    public WaterLevelRecipeSerializer() {
    }

    public T fromJson(ResourceLocation recipeId, JsonObject json) {
        /* Default if values not be written
        * nbt null
        * fluid null
        * mobEffects List(Empty)
        * ingredient Ingredient.EMPTY
        * waterLevel 0
        * waterSaturationLevel 0
        * */
        int waterLevel = 0;
        int waterSaturationLevel = 0;
        Fluid fluid = null;
        CompoundTag compoundTag = null;
        List<MobEffectInstance> effectInstances = new ArrayList<>();

        String group = GsonHelper.getAsString(json, "group", "");

        //ingredient
        Ingredient ingredient = Ingredient.EMPTY;
        if (GsonHelper.isValidNode(json,"ingredient")) {
            JsonElement jsonelement = GsonHelper.isArrayNode(json, "ingredient") ? GsonHelper.getAsJsonArray(json, "ingredient") : GsonHelper.getAsJsonObject(json, "ingredient");
            ingredient = Ingredient.fromJson(jsonelement);
        }


        //mobEffects
        if (GsonHelper.isArrayNode(json ,"mob_effects")){
            JsonArray effectsJsonArray = GsonHelper.getAsJsonArray(json, "mob_effects");
            for (JsonElement effect : effectsJsonArray) {
                JsonObject mobEffectJsonObj = effect.getAsJsonObject();
                int duration = GsonHelper.getAsInt(mobEffectJsonObj, "duration");
                int amplifier = GsonHelper.getAsInt(mobEffectJsonObj, "amplifier");
                String name = GsonHelper.getAsString(mobEffectJsonObj, "name");
                if (duration > 0 && amplifier >= 0) {
                    MobEffect mobEffect = ForgeRegistries.MOB_EFFECTS.getValue(ResourceLocation.tryParse(name));
                    if (mobEffect != null) {
                        effectInstances.add(new MobEffectInstance(mobEffect, duration, amplifier));
                    }
                }
            }
        }


        //nbt
        if (GsonHelper.isValidNode(json ,"nbt")) {
            JsonObject nbt = GsonHelper.getAsJsonObject(json, "nbt");
            try {
                compoundTag = NbtUtils.snbtToStructure(nbt.toString());
            } catch (CommandSyntaxException e) {
                System.out.println(recipeId + ": no nbt.");
            }
            //todo debug
        }

        //fluid
        if (GsonHelper.isValidNode(json ,"fluid")){
            String fluidName = GsonHelper.getAsString(json, "fluid", "");
            fluid = ForgeRegistries.FLUIDS.getValue(new ResourceLocation(fluidName));
        }

        //water level
        waterLevel = GsonHelper.getAsInt(json, "water_level", 0);
        waterSaturationLevel = GsonHelper.getAsInt(json, "water_saturation_level", 0);

        return (T) new WaterLevelAndEffectRecipe(recipeId, group, ingredient, waterLevel, waterSaturationLevel, effectInstances, fluid, compoundTag);
    }

    @Override
    public T fromNetwork(ResourceLocation recipeId, FriendlyByteBuf packetBuffer) {
        List<MobEffectInstance> mobEffectInstances = new ArrayList<>();
        String group = packetBuffer.readUtf();//1
        Ingredient ingredient = Ingredient.fromNetwork(packetBuffer);//2
        int waterLevel = packetBuffer.readVarInt();//3
        int waterSaturationLevel = packetBuffer.readVarInt();//4
        Fluid fluid = ForgeRegistries.FLUIDS.getValue(new ResourceLocation(packetBuffer.readUtf()));//5
        CompoundTag compoundTag = packetBuffer.readNbt();//6
        for (int i = 0; i < packetBuffer.readByte(); i++) {
            String mobEffectName = packetBuffer.readUtf();
            int duration = packetBuffer.readVarInt();
            int amplifier = packetBuffer.readVarInt();

            MobEffect mobEffect = ForgeRegistries.MOB_EFFECTS.getValue(ResourceLocation.tryParse(mobEffectName));
            mobEffectInstances.add(new MobEffectInstance(mobEffect, duration, amplifier));
        }
        return (T) new WaterLevelAndEffectRecipe(recipeId, group, ingredient, waterLevel, waterSaturationLevel, mobEffectInstances, fluid, compoundTag);
    }

    @Override
    public void toNetwork(FriendlyByteBuf buffer, T recipe) {
        buffer.writeUtf(recipe.getGroup());
        recipe.getIngredient().toNetwork(buffer);
        buffer.writeVarInt(recipe.getWaterLevel());
        buffer.writeVarInt(recipe.getWaterSaturationLevel());
        buffer.writeUtf(recipe.getFluid() == null ? "" : recipe.getFluid().getRegistryName().toString());
        buffer.writeNbt(recipe.getCompoundTag());
        buffer.writeByte(recipe.getMobEffectInstances().size());
        for (MobEffectInstance mobEffectInstance : recipe.getMobEffectInstances()) {
            buffer.writeUtf(mobEffectInstance.getEffect().getRegistryName().toString());
            buffer.writeVarInt(mobEffectInstance.getDuration());
            buffer.writeVarInt(mobEffectInstance.getAmplifier());
        }
    }
}

File: src/main/kotlin/watersource/world/recipe/serializer/WaterLevelRecipeSerializer.kt
package watersource.world.recipe.serializer

import com.google.gson.JsonObject
import net.minecraft.core.Registry
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.TagParser
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.GsonHelper
import net.minecraft.world.item.Item
import net.minecraft.world.item.crafting.RecipeSerializer
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.registries.ForgeRegistries
import watersource.world.recipe.type.WaterLevelRecipe
import watersource.world.recipe.type.WaterLevelRecipeSort
import watersource.world.recipe.type.WaterLevelRecipeSort.*

class WaterLevelRecipeSerializer : RecipeSerializer<WaterLevelRecipe> {
    object KEYS{
        const val sortKey = "sort"
        const val groupKey = "group"
        const val fluidKey = "fluid"
        const val itemKey = "item"
        const val nameKey = "name"
        const val tagKey = "tag"
        const val nbtKey = "nbt"
        const val waterKey = "water"
        const val saturationModifierKey = "saturationModifier"
    }
    override fun equals(other: Any?): Boolean {
        return super.equals(other)
    }

    override fun hashCode(): Int {
        return super.hashCode()
    }

    override fun toString(): String {
        return super.toString()
    }

    override fun fromJson(resourceLocation: ResourceLocation, jsonObject: JsonObject): WaterLevelRecipe {
        val recipe : WaterLevelRecipe = when {
            GsonHelper.isObjectNode(jsonObject, KEYS.fluidKey) -> {
                val fluidJson: JsonObject = GsonHelper.getAsJsonObject(jsonObject, KEYS.fluidKey)
                val name = GsonHelper.getAsString(fluidJson, KEYS.nameKey)
                val tag = GsonHelper.getAsString(fluidJson, KEYS.tagKey)
                val water = GsonHelper.getAsFloat(fluidJson, KEYS.waterKey)
                val saturationModifier = GsonHelper.getAsFloat(fluidJson, KEYS.saturationModifierKey)

                val fluid = ForgeRegistries.FLUIDS.getValue(ResourceLocation(name))
                WaterLevelRecipe.createFluidRecipe(fluid, tag, water, saturationModifier)
            }
            GsonHelper.isObjectNode(jsonObject, KEYS.itemKey) -> {
                val itemJson: JsonObject = GsonHelper.getAsJsonObject(jsonObject, KEYS.itemKey)
                val name = GsonHelper.getAsString(itemJson, KEYS.nameKey)
                val tag = GsonHelper.getAsString(itemJson, KEYS.tagKey)
                val nbtString = GsonHelper.getAsString(itemJson, KEYS.nbtKey)
                val water = GsonHelper.getAsFloat(itemJson, KEYS.waterKey)
                val saturationModifier = GsonHelper.getAsFloat(itemJson, KEYS.saturationModifierKey)

                val item = ForgeRegistries.ITEMS.getValue(ResourceLocation(name))
                val nbt = TagParser.parseTag(nbtString)
                WaterLevelRecipe.createItemRecipe(item, tag, water, saturationModifier, nbt)
            }
            else -> WaterLevelRecipe()
        }
        recipe.recipeGroup = GsonHelper.getAsString(jsonObject, KEYS.groupKey)
        return recipe
    }

    override fun fromNetwork(resourceLocation: ResourceLocation, buf: FriendlyByteBuf): WaterLevelRecipe {
        val group = buf.readUtf()
        val sort = WaterLevelRecipeSort.values()[buf.readShort().toInt()]
        val water: Float = buf.readFloat()
        val saturationModifier: Float = buf.readFloat()

        var recipe: WaterLevelRecipe =
            when (sort) {
                ITEM -> {
                    val item: Item? = buf.readById(Registry.ITEM)
                    val tag: String = buf.readUtf()
                    val nbt: CompoundTag? = buf.readNbt()
                    WaterLevelRecipe.createItemRecipe(item, tag, water, saturationModifier, nbt)
                }

                FLUID -> {
                    val fluid: Fluid? = buf.readById(Registry.FLUID)
                    val tag: String = buf.readUtf()
                    WaterLevelRecipe.createFluidRecipe(fluid, tag, water, saturationModifier)
                }
            }
        recipe.recipeGroup = group
        return recipe;
    }

    override fun toNetwork(buf: FriendlyByteBuf, recipe: WaterLevelRecipe) {
        buf.writeUtf(recipe.recipeGroup)
        buf.writeShort(recipe.sort.ordinal)
        buf.writeFloat(recipe.water)
        buf.writeFloat(recipe.saturationModifier)
        when (recipe.sort) {
            ITEM -> {
                buf.writeId(Registry.ITEM, recipe.item)
                buf.writeUtf(recipe.tag)
                buf.writeNbt(recipe.nbt)
            }

            FLUID -> {
                buf.writeId(Registry.FLUID, recipe.fluid)
                buf.writeUtf(recipe.tag)
            }
        }
    }
}